modules/core/src/main/java/org/apache/ignite/internal/util/GridHandleTable.java

```diff
@@ -181,6 +203,32 @@ private void growEntries() {
         objs = newObjs;
     }
 
+    /**
+     * Tries to gradually shrink hash table by factor of two when it's cleared.
+     *
+     * @return {@code true} if shrinked the table, {@code false} otherwise.
+     */
+    private boolean shrink() {
+        int newLen = objs.length;
+
+        if (newLen > initCap) {
+            int shrinked = (newLen - 1) / 2;
```

Comment:
why `newLen - 1`?

Reply:
I think this code can actually be written a little bit simpler:
```
int newLen = Math.max(objs.length / 2, initCap);

if (newLen >= size && newLen < objs.length) {
    init(newLen);

    return true;
}

return false;
```

Reply:
Because we grow it as "(size  << 1) + 1"

Reply:
And why do we do that?